lib/sqlalchemy/sql/type_api.py

```diff
@@ -470,13 +470,65 @@ def _generic_type_affinity(self):
                     "sqlalchemy.sql.sqltypes",
                     "sqlalchemy.sql.type_api",
                 )
-                and t._is_generic_type()
+                and issubclass(t, TypeEngine)
+                and t is not TypeEngine
+                and t.__name__[0] != "_"
             ):
-                if t in (TypeEngine, UserDefinedType):
-                    return NULLTYPE.__class__
-                return t
+                if t.__name__.isupper() and not best_uppercase:
+                    best_uppercase = t
+                elif not t.__name__.isupper() and not best_camelcase:
+                    best_camelcase = t
+
+        return best_camelcase or best_uppercase or NULLTYPE.__class__
+
+    def as_generic(self, allow_nulltype=False):
+        """
+        Return an instance of the generic type corresponding to this type
+        using heuristic rule. The method may be overridden if this
+        heuristic rule is not sufficient.
+
+        >>> from sqlalchemy.dialects.mysql import INTEGER
+        >>> INTEGER(display_width=4).as_generic()
+        Integer()
+
+        >>> from sqlalchemy.dialects.mysql import NVARCHAR
+        >>> NVARCHAR(length=100).as_generic()
+        Unicode(length=100)
+
+        .. versionadded:: 1.4.0b2
+        """
+        from sqlalchemy import Enum
+
+        if isinstance(self, Enum):
+            if hasattr(self, "enums"):
+                args = self.enums
+            else:
+                raise NotImplementedError(
+                    "TypeEngine.as_generic() heuristic "
+                    "is undefined for types that inherit Enum but do not have "
+                    "an `enums` attribute."
+                )
         else:
-            return self.__class__
+            args = ()
+
+        if (
+            not allow_nulltype
+            and self._generic_type_affinity == NULLTYPE.__class__
+        ):
+            raise NotImplementedError(
+                "Default TypeEngine.as_generic() "
+                "heuristic method was unsuccessful for {}. A custom "
+                "as_generic() method must be implemented for this "
+                "type class.".format(
+                    self.__class__.__module__ + "." + self.__class__.__name__
+                )
+            )
+
+        return util.constructor_copy(self, self._generic_type_affinity, *args)
+
+    @classmethod
+    def _uses_as_generic_heuristic(cls):
+        return cls.as_generic == TypeEngine.as_generic
```

Comment:
**mike bayer** (zzzeek) wrote:

seems to work, surprised.  we have similar kinds of things in TypeDecorator and they look like this:

    @util.memoized_property
    def _has_bind_processor(self):

        return (
            self.__class__.process_bind_param.__code__
            is not TypeDecorator.process_bind_param.__code__
        )

*checks*

hmmm yes the way it is here works also.   perhaps the __code__ approach was only needed for older Python versions.  (pre 2.7).

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2375